File: server/service/article.go
package service

import (
	"context"
	"server/model/elasticsearch"
	"server/model/other"
	"server/model/request"
	"server/utils"

	"github.com/elastic/go-elasticsearch/v8/typedapi/eql/search"
	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
	"github.com/elastic/go-elasticsearch/v8/typedapi/types/enums/sortorder"
)

type ArticleService struct {
}

func (articleService *ArticleService) ArticleInfoByID(id string) (elasticsearch.Article, error) {
	go func() {
		articleViews := articleService.NewArticleView()
		articleViews.Set(id)
	}()
	return articleService.Get(id)
}
func (articleService *ArticleService) ArticleSearch(info request.ArticleSearch) (interface{}, int64, error) {
	req := &search.Request{
		Query: &types.Query{},
	}

	boolQuery := &types.BoolQuery{}

	if info.Query != "" {
		boolQuery.Should = []types.Query{
			{Match: map[string]types.MatchQuery{"title": {Query: info.Query}}},
			{Match: map[string]types.MatchQuery{"keyword": {Query: info.Query}}},
			{Match: map[string]types.MatchQuery{"abstract": {Query: info.Query}}},
			{Match: map[string]types.MatchQuery{"content": {Query: info.Query}}},
		}
	}

	if info.Tag != "" {
		boolQuery.Must = []types.Query{
			{Match: map[string]types.MatchQuery{"tags": {Query: info.Tag}}},
		}
	}

	if info.Category != "" {
		boolQuery.Filter = []types.Query{
			{Term: map[string]types.TermQuery{"category": {Value: info.Category}}},
		}
	}

	if boolQuery.Should != nil || boolQuery.Must != nil || boolQuery.Filter != nil {
		req.Query.Bool = boolQuery
	} else {
		req.Query.MatchAll = &types.MatchAllQuery{}
	}

	if info.Sort != "" {
		var sortField string
		switch info.Sort {
		case "time":
			sortField = "created_at"
		case "view":
			sortField = "views"
		case "comment":
			sortField = "comments"
		case "like":
			sortField = "likes"
		default:
			sortField = "created_at"
		}

		var order sortorder.SortOrder
		if info.Order != "asc" {
			order = sortorder.Desc
		} else {
			order = sortorder.Asc
		}

		req.Sort = []types.SortCombinations{
			types.SortOptions{
				SortOptions: map[string]types.FieldSort{
					sortField: {Order: &order},
				},
			},
		}
	}

	option := other.EsOption{
		PageInfo:       info.PageInfo,
		Index:          elasticsearch.ArticleIndex(),
		Request:        req,
		SourceIncludes: []string{"created_at", "cover", "title", "abstract", "category", "tags", "views", "comments", "likes"},
	}

	return utils.EsPagination(context.TODO(), option)
}
func (articleService *ArticleService) ArticleCategory() {

}
func (articleService *ArticleService) ArticleTags() {

}
func (articleService *ArticleService) ArticleLike() {

}
func (articleService *ArticleService) ArticleIsLike() {

}
func (articleService *ArticleService) ArticleLikesList() {

}
func (articleService *ArticleService) ArticleCreate() {

}
func (articleService *ArticleService) ArticleDelete() {

}
func (articleService *ArticleService) ArticleUpdate() {

}
func (articleService *ArticleService) ArticleList() {

}

File: web/src/api/article.ts
import type {Hit, PageInfo, PageResult} from "@/api/common";
import type {ApiResponse} from "@/utils/request";
import service from "@/utils/request";

export interface Article {
    created_at: string;
    updated_at: string;

    cover: string;
    title: string;
    keyword: string;
    category: string;
    tags: string[];
    abstract: string;
    content: string;

    views: number;
    comments: number;
    likes: number;
}

export interface ArticleLikeRequest {
    article_id: string;
}

export const articleLike = (data: ArticleLikeRequest): Promise<ApiResponse<undefined>> => {
    return service({
        url: '/article/like',
        method: 'post',
        data: data
    })
}

export const articleIsLike = (data: ArticleLikeRequest): Promise<ApiResponse<boolean>> => {
    return service({
        url: '/article/isLike',
        method: 'get',
        params: data
    })
}

export const articleLikesList = (data: PageInfo): Promise<ApiResponse<PageResult<Hit<Article>>>> => {
    return service({
        url: '/article/likesList',
        method: 'get',
        params: data
    })
}

export interface ArticleCreateRequest {
    cover: string;
    title: string;
    category: string;
    tags: string[];
    abstract: string;
    content: string;
}

export const articleCreate = (data: ArticleCreateRequest): Promise<ApiResponse<undefined>> => {
    return service({
        url: '/article/create',
        method: 'post',
        data: data
    });
}

export interface ArticleDeleteRequest {
    ids: string[];
}

export const articleDelete = (data: ArticleDeleteRequest): Promise<ApiResponse<undefined>> => {
    return service({
        url: '/article/delete',
        method: 'delete',
        data: data
    });
}

export interface ArticleUpdateRequest {
    id: string;
    cover: string;
    title: string;
    category: string;
    tags: string[];
    abstract: string;
    content: string;
}

export const articleUpdate = (data: ArticleUpdateRequest): Promise<ApiResponse<undefined>> => {
    return service({
        url: '/article/update',
        method: 'put',
        data: data
    });
}

export interface ArticleListRequest extends PageInfo {
    title: string | null;
    category: string | null;
    abstract: string | null;
}

export const articleList = (data: ArticleListRequest): Promise<ApiResponse<PageResult<Hit<Article>>>> => {
    return service({
        url: '/article/list',
        method: 'get',
        params: data,
    });
}

export const articleInfoByID = (id: string): Promise<ApiResponse<Article>> => {
    return service({
        url: '/article/'+id,
        method: 'get',
    });
}

export interface ArticleSearchRequest extends PageInfo {
    query: string;
    category: string;
    tag: string;
    sort: string;
    order: string;
}

export const articleSearch = (data: ArticleSearchRequest): Promise<ApiResponse<PageResult<Hit<Article>>>> => {
    return service({
        url: '/article/search',
        method: 'get',
        params: data,
    });
}

export interface ArticleCategory {
    category: string;
    number: number;
}

export const articleCategory = (): Promise<ApiResponse<ArticleCategory[]>> => {
    return service({
        url: '/article/category',
        method: 'get',
    });
}

export interface ArticleTag {
    tag: string;
    number: number;
}

export const articleTags = (): Promise<ApiResponse<ArticleTag[]>> => {
    return service({
        url: '/article/tags',
        method: 'get',
    });
}
